examples/step-66/step-66.cc

```diff
@@ -0,0 +1,1156 @@
+  void GelfandProblem<dim>::assemble_rhs()
+  {
+    TimerOutput::Scope t(computing_timer, "assemble right hand side");
+
+    residual_operator.apply(system_rhs, solution);
+
+    system_rhs *= -1.0;
+  }
+
+
+
+  // @sect4{GelfandProblem::compute_residual}
+
+  // With the help of the ResidualOperator the next function computes the
+  // residual of the nonlinear problem. The input argument alpha becomes
+  // important if we would use an adaptive version of the Newton method. Then
+  // for example we would compute the residual for different step lengths and
+  // compare the residuals. However for our problem the full Newton step with
+  // $\alpha=1$ is the best we can do. An adaptive verison of Newton's method
+  // becomes interesting if we have no good initial value. Note, the theory
+  // states that Newton's method converges with quadratic order, but only if we
+  // have an appropriate initial value. For wrong initial values the Newton
+  // method diverges even with quadratic order. A common way is then to use a
+  // damped version $\alpha<1$ until the Newton step is good enough and the full
+  // Newton step can be performed. This was also discussed in step-15.
+  template <int dim>
+  double GelfandProblem<dim>::compute_residual(const double alpha)
+  {
+    TimerOutput::Scope t(computing_timer, "compute residual");
+
+    LinearAlgebra::distributed::Vector<double> residual;
+    LinearAlgebra::distributed::Vector<double> evaluation_point;
+
+    system_matrix.initialize_dof_vector(residual);
+    system_matrix.initialize_dof_vector(evaluation_point);
+
+    evaluation_point = solution;
+    if (alpha > 1e-12)
+      evaluation_point.add(alpha, newton_update);
+
+    residual_operator.apply(residual, evaluation_point);
+
+    return residual.l2_norm();
+  }
+
+
+
+  // @sect4{GelfandProblem::compute_update}
+
+  // In order to compute the Newton updates in each step Newton step we solve
+  // the linear system with the cg algorithm together with a geometric multigrid
+  // preconditioner. For this we first setup the PreconditionMG object with a
+  // Chebyshev smoother like we did in step-37.
+  template <int dim>
+  void GelfandProblem<dim>::compute_update()
+  {
+    TimerOutput::Scope t(computing_timer, "compute update");
+
+    // We remember that the Jacobian depends on the last Newton step stored in
+    // the solution vector. So we update the ghost values of the Newton step and
+    // pass it to the JacobianOperator to store the information.
+    solution.update_ghost_values();
+
+    system_matrix.evaluate_newton_step(solution);
+
+
+    // Next we have to pass the last Newton step also to the multilevel
+    // operators. However, therefore we need to interpolate the Newton step to
+    // all levels of the triangulation. This is done with the interpolate_to_mg
+    // function of the MGTransferMatrixFree class.
+    MGTransferMatrixFree<dim, float> mg_transfer(mg_constrained_dofs);
+    mg_transfer.build(dof_handler);
```

Comment:
Can you make `MGTransferMatrixFree` a class member? From an educational point of view, I think it would make sense to show that the transfer only needs to be set up once and can be re-used in subsequent nonlinear iterations. This is different to the `mg_smoother` object that must indeed be rebuilt because the diagonal changes (even though one could try to find a balance of not recomputing in every iteration - that would make for a good possibility for extensions, though).

Reply:
Thats a good point! But maybe I didnt get the idea for the 'possibility of extensions'.